### PR TITLE
Separate metadata into its own table for Fact Tables

### PR DESCRIPTION
**Pull request by agavra**
TWCS doesn't perform well if there are rows that are replaced. This assumption was not respected with the metadata (offset) rows, that were replaced regularly potentially messing up compaction.

The only meaningful change is in `kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java`

### CODE REVIEW

File: kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java [hunk 1]
```diff
@@ -76,30 +73,43 @@ public CassandraFactSchema(final CassandraClient client) {
 
   @Override
   @CheckReturnValue
-  public SimpleStatement create(final String tableName, Optional<Duration> ttl) {
+  public void create(final String tableName, Optional<Duration> ttl) {
     LOG.info("Creating fact data table {} in remote store.", tableName);
-    final CreateTable createTable = SchemaBuilder
+    CreateTableWithOptions createTable = SchemaBuilder
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:78]: nit: maybe this is paranoid but I'm a bit nervous about non-final variables like this...could we maybe wrap the shared options in a `createTable()` method and have something like
```
final CreateTable createTable;
if (ttl.isPresent()) {
 createTable = createTable().withTTl(...).withCompaction(...);
} else {
  createTable = createTable().withCompaction(...);
}
```

File: kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java [hunk 2]
```diff
@@ -76,30 +73,43 @@ public CassandraFactSchema(final CassandraClient client) {
...
+    // separate metadata from the main table for the fact schema, this is acceptable
+    // because we don't use the metadata at all for fencing operations and writes to
+    // it do not need to be atomic (transactional with the original table). we cannot
+    // effectively use the same table (as we do with the normal KeyValueSchema) because
+    // TWCS cannot properly compact files if there are any overwrites, which there are
+    // for the metadata columns
+    final CreateTableWithOptions createMetadataTable = SchemaBuilder
+        .createTable(metadataTable(tableName))
+        .ifNotExists()
+        .withPartitionKey(PARTITION_KEY.column(), DataTypes.INT)
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:108]: Are we going from a TINYINT to an INT here? Is that intentional?

File: kafka-client/src/main/java/dev/responsive/db/RemoteSchema.java [hunk 3]
```diff
@@ -34,7 +33,7 @@ public interface RemoteSchema<K> {
    * desired schema.
    */
   @CheckReturnValue
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/db/RemoteSchema.java:35]: this should probably be removed, hm?

File: kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java [hunk 4]
```diff
@@ -308,11 +313,9 @@ public KeyValueIterator<Bytes, byte[]> all(
...
   }
 
-  private static ByteBuffer metadataKey(final int partition) {
-    final ByteBuffer key = ByteBuffer.wrap(new byte[METADATA_KEY.get().length + Integer.BYTES]);
-    key.put(METADATA_KEY.get());
-    key.putInt(partition);
-    return key.position(0);
+  // Visible for Testing
+  static String metadataTable(final String tableName) {
+    return tableName + "_md";
```
Comment (ableegoldman) [inline kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:318]: This should never change, right? Could we make this private and then hard-code this expectation in the tests to verify this?

Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:318]: yeah we could do that

File: kafka-client/src/main/java/dev/responsive/db/RemoteSchema.java (hunk 3 above)
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/db/RemoteSchema.java:35]: oops thanks!

File: kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java (hunk 2 above)
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:108]: yes, TINYINT was for the "row type", the partition key on the other hand is the Kafka Partition which is technically an integer type.

**Review by rodesai — COMMENTED** [review]
LGTM. One question inline

File: kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java (hunk 2 above)
Comment (agavra) [reply kafka-client/src/main/java/dev/responsive/db/CassandraFactSchema.java:103]: 👍

**Pull request merged**